client/forest-manager/src/rocksdb.rs

```diff
@@ -0,0 +1,611 @@
+            matches!(proof.proven.first().expect("Proven leaves should have proven 1 challenge"), Proven::ExactKey(leaf) if leaf.key.as_ref() == lookup_key)
+        );
+    }
+
+    #[test]
+    fn test_generate_proof_neighboor_keys() {
+        let mut forest_storage = setup_storage::<LayoutV1<RefHasher>>();
+
+        let (lookup_key1, _) =
+            create_and_insert_metadata(&mut forest_storage, "Bob", vec![7, 8, 9], 200);
+        let (lookup_key2, _) =
+            create_and_insert_metadata(&mut forest_storage, "Alice", vec![7, 8, 10], 200);
+
+        let smallest = std::cmp::min(lookup_key1, lookup_key2);
+        let largest = std::cmp::max(lookup_key1, lookup_key2);
+        let mut challenge = smallest.clone();
+        challenge[0] = challenge[0] + 1;
+
+        let proof = forest_storage.generate_proof(&vec![challenge]).unwrap();
+
+        assert_eq!(proof.proven.len(), 1);
+        assert!(
+            matches!(proof.proven.first().expect("Proven leaves should have proven 1 challenge"), Proven::NeighbourKeys((Some(left_leaf), Some(right_leaf))) if left_leaf.key.as_ref() == smallest && right_leaf.key.as_ref() == largest)
+        );
+    }
+
+    // TODO: Fix this test
+    #[test]
+    #[ignore = "double ended iterator has inconsistent behaviour"]
+    fn test_generate_proof_challenge_before_first_leaf() {
+        let mut forest_storage = setup_storage::<LayoutV1<RefHasher>>();
+
+        let (lookup_key1, _) =
+            create_and_insert_metadata(&mut forest_storage, "Alice", vec![10], 200);
+        let (_lookup_key2, _) =
+            create_and_insert_metadata(&mut forest_storage, "Alice", vec![11], 200);
+
+        let mut challenge = lookup_key1.clone();
+        challenge[0] = challenge[0] - 1;
+
+        let proof = forest_storage.generate_proof(&vec![challenge]).unwrap();
+
+        assert_eq!(proof.proven.len(), 1);
+        assert!(
+            matches!(proof.proven.first().expect("Proven leaves should have proven 1 challenge"), Proven::NeighbourKeys((None, Some(leaf))) if leaf.key.as_ref() == lookup_key1)
+        );
+    }
+
+    #[test]
+    fn test_generate_proof_challenge_after_last_leaf() {
+        let mut forest_storage = setup_storage::<LayoutV1<RefHasher>>();
+
+        let (lookup_key1, _) =
+            create_and_insert_metadata(&mut forest_storage, "Alice", vec![1, 2, 3], 200);
+        let (lookup_key2, _) =
+            create_and_insert_metadata(&mut forest_storage, "Bob", vec![7, 8, 9], 200);
+
+        let largest = std::cmp::max(lookup_key1, lookup_key2);
+        let mut challenge = largest.clone();
+        challenge[0] = challenge[0] + 1;
+
+        let proof = forest_storage.generate_proof(&vec![challenge]).unwrap();
+
+        assert_eq!(proof.proven.len(), 1);
+        assert!(
+            matches!(proof.proven.first().expect("Proven leaves should have proven 1 challenge"), Proven::NeighbourKeys((Some(leaf), None)) if leaf.key.as_ref() == largest)
+        );
+    }
+}
```

Comment:
I would add a test with more than 16 data insertions to be absolutely sure that there are intermediate nodes in the trie, and test how this codabase handles intermediate nodes. If it updates and clears them accordingly, for example.

Reply:
Agreed